exercises/practice/sgf-parsing/.meta/example.ex

```diff
@@ -0,0 +1,207 @@
+      satisfy(p)
+    ])
+  end
+
+  def escape("\\", "n"), do: "\n"
+  def escape("\\", "t"), do: "\t"
+  def escape("\\", "]"), do: "]"
+  def escape("\\", "["), do: "["
+
+  # PARSER COMBINATORS LIBRARY
+  # Inspired from Haskell libraries like Parsec
+  # and https://serokell.io/blog/parser-combinators-in-elixir
+
+  def run_parser(parser, string), do: parser.(string)
+
+  def eof(parser) do
+    fn string ->
+      with {:ok, _, ""} = ok <- parser.(string) do
+        ok
+      else
+        {:ok, _a, rest} -> {:error, "Not end of file", rest}
+        err -> err
+      end
+    end
+  end
+
+  def satisfy(p) do
+    fn
+      <<char, rest::bitstring>> = string ->
+        if p.(char) do
+          {:ok, <<char>>, rest}
+        else
+          {:error, "unexpected #{char}", string}
+        end
+
+      "" ->
+        {:error, "unexpected end of string", ""}
+    end
+  end
+
+  def char(c), do: satisfy(&(&1 == c)) |> error("expected character #{<<c>>}")
+
+  def string(str) do
+    str
+    |> to_charlist
+    |> Enum.map(&char/1)
+    |> Enum.reduce(inject(""), &lift2(fn a, b -> a <> b end, &1, &2))
+  end
+
+  def some(parser) do
+    fn input ->
+      with {:ok, result, rest} <- parser.(input),
+           {:ok, results, rest} <- many(parser).(rest) do
+        {:ok, [result | results], rest}
+      end
+    end
+  end
+
+  def many(parser) do
+    fn input ->
+      with {:ok, result, rest} <- some(parser).(input) do
+        {:ok, result, rest}
+      else
+        {:error, _err, ^input} -> {:ok, [], input}
+        err -> err
+      end
+    end
+  end
+
+  def one_of(parsers) when is_list(parsers) do
+    fn string ->
+      Enum.reduce_while(parsers, {:error, "no parsers", string}, fn
+        _parser, {:ok, _, _} = result -> {:halt, result}
+        parser, _err -> {:cont, parser.(string)}
+      end)
+    end
+  end
+
+  def map(parser, f) do
+    fn string ->
+      with {:ok, a, rest} <- parser.(string) do
+        {:ok, f.(a), rest}
+      end
+    end
+  end
+
+  def error(parser, err) do
+    fn string ->
+      with {:error, _err, rest} <- parser.(string) do
+        {:error, err, rest}
+      end
+    end
+  end
+
+  def drop(p1, p2) do
+    fn string ->
+      with {:ok, a, rest} <- p1.(string),
+           {:ok, _, rest} <- p2.(rest) do
+        {:ok, a, rest}
+      end
+    end
+  end
+
+  def drop_and(p1, p2) do
+    fn string ->
+      with {:ok, _, rest} <- p1.(string) do
+        p2.(rest)
+      end
+    end
+  end
+
+  def inject(a) do
+    fn string -> {:ok, a, string} end
+  end
+
+  def lift2(pair, p1, p2) do
+    fn string ->
+      with {:ok, a, rest} <- p1.(string),
+           {:ok, b, rest} <- p2.(rest) do
+        {:ok, pair.(a, b), rest}
+      end
+    end
+  end
+
+  def validate(parser, p, err) do
+    fn string ->
+      with {:ok, result, rest} <- parser.(string) do
+        if p.(result) do
+          {:ok, result, rest}
+        else
+          {:error, err, rest}
+        end
+      end
```

Comment:
When there is a single expression in the `with`, is there a benefit to using `with` over

```suggestion
      {:ok, result, rest} = parser.(string)
      
      if p.(result) do
        {:ok, result, rest}
      else
        {:error, err, rest}
      end
```

Reply:
The `with` here is important, because I need the error to be passed further if `parser.(string)` fails. If I use ` {:ok, result, rest} = parser.(string)` and the parser fails, I will get an exception.

Reply:
:+1: